pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,448 @@
+package symbolizer
+
+import (
+	"bytes"
+	"context"
+	"debug/elf"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"path/filepath"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	pprof "github.com/google/pprof/profile"
+	"github.com/prometheus/client_golang/prometheus"
+
+	googlev1 "github.com/grafana/pyroscope/api/gen/proto/go/google/v1"
+	"github.com/grafana/pyroscope/lidia"
+	"github.com/grafana/pyroscope/pkg/objstore"
+)
+
+type DebuginfodClient interface {
+	FetchDebuginfo(ctx context.Context, buildID string) (io.ReadCloser, error)
+}
+
+type Config struct {
+	Enabled                  bool                 `yaml:"enabled"`
+	DebuginfodURL            string               `yaml:"debuginfod_url"`
+	PersistentDebugInfoStore DebugInfoStoreConfig `yaml:"persistent_debuginfo_store"`
+}
+
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enabled, "symbolizer.enabled", false, "Enable symbolization for unsymbolized profiles")
+	f.StringVar(&cfg.DebuginfodURL, "symbolizer.debuginfod-url", "https://debuginfod.elfutils.org", "URL of the debuginfod server")
+	cfg.PersistentDebugInfoStore.Storage.RegisterFlagsWithPrefix("symbolizer.persistent-debuginfo-store.storage.", f)
+}
+
+type Symbolizer struct {
+	logger  log.Logger
+	client  DebuginfodClient
+	store   DebugInfoStore
+	metrics *metrics
+}
+
+func New(logger log.Logger, cfg Config, reg prometheus.Registerer, bucket objstore.Bucket) (*Symbolizer, error) {
+	metrics := newMetrics(reg)
+
+	store := NewObjstoreDebugInfoStore(bucket)
+
+	client, err := NewDebuginfodClient(logger, cfg.DebuginfodURL, metrics)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Symbolizer{
+		logger:  logger,
+		client:  client,
+		store:   store,
+		metrics: metrics,
+	}, nil
+}
+
+func (s *Symbolizer) SymbolizePprof(ctx context.Context, profile *googlev1.Profile) error {
+	start := time.Now()
+	status := statusSuccess
+	defer func() {
+		s.metrics.profileSymbolization.WithLabelValues(status).Observe(time.Since(start).Seconds())
+	}()
+
+	mappingsToSymbolize := make(map[uint64]bool)
+	for i, mapping := range profile.Mapping {
+		if mapping.HasFunctions && mapping.HasFilenames && mapping.HasLineNumbers {
+			continue
+		}
+		mappingsToSymbolize[uint64(i+1)] = true
+	}
+	if len(mappingsToSymbolize) == 0 {
+		return nil
+	}
+
+	locationsByMapping, err := s.groupLocationsByMapping(profile, mappingsToSymbolize)
+	if err != nil {
+		return fmt.Errorf("grouping locations by mapping: %w", err)
+	}
+
+	for mappingID, locations := range locationsByMapping {
+		if err := s.symbolizeLocationsForMapping(ctx, profile, mappingID, locations); err != nil {
```

Comment:
I suggest we do not fail the query if there are any errors during symbolization. Let's treat all the errors the same - createNotFoundSymbolz